### DISCUSSION TITLE
build and read arrays in struct

### DISCUSSION THREAD

**Discussion by michbens** [dvidelabs/flatcc#230] [Q&A] [answered]
Hello,
I heave the following schema and I cannot find a way to build MyTable with payload as Struct2. 
If someone can help.
Thanks
```
table Table1
{
...
}

struct Struct2
{
    field: uint32;
    array: [uint32:10];
}

union Payload {Table1, Struct2}

table MyTable
{
    payload: Payload;
}
```
The problem is to initiate the array 
I've tried this but it doesn't work. I think it is because it's a struct and not a table
```
uint32_t field = 3;
uint32_t a_regular_array[10] = {...};
flatbuffers_uint32_vec_ref_t array;

array = flatbuffers_uint32_vec_create(&builder, a_regular_array, 10);

ns_Payload_union_ref_t pl = ns_Payload_as_Struct2(ns_Struct2_create(&builder, field, array));
ns_MyTable_ref_t tbl = Cfg_Configuration_create(&builder, pl);
```

I also tried this with no success

```
ns_Struct2_t *st;

ns_MyTable_start_as_root(&builder);
st = ns_MyTable_Payload_start(&builder);
st->field = 3;
for (int i = 0 ; i < 10 ; i++) {
    st->array[ן] = a_regular_array[i];
}
ns_MyTable_Payload_end(&builder);
ns_MyTable_end_as_root(&builder);
```

**Comment by mikkelfj** [comment]
The first attempt is because it is a struct as you suspect. Structs are selfcontained objects without external references and fixed length arrays in structs are not FlatBuffer vectors.

Your second attempt is better. I don't recall the exact syntax but it seems like you are are starting the payload field as a generic Payload union instead of a specific instance of type "Struct2".

It would be more along the lines of `ns_MyTable_Payload_Struct2_start` instead of `ns_MyTable_Payload_start`.
More info here:
https://github.com/dvidelabs/flatcc/blob/master/doc/builder.md#unions-of-strings-and-structs

Also worth noting: normally a struct is stored in the same memory blob as other table fields such as integers, but when it is a union it is allocated separately. The union is two fields: a pointer (offset), and a type. You need to make sure the pointer to the struct is allcoted, and that the type is set. Some of the helper functions do some of that work automatically.

Please let me know if this is enough to get you going, or I can dig deeper.

**Reply by michbens** [reply]
Thanks a lot. This one helped.
I've done the following (as for table but just passing native array instead of flatbuffers_uint32_vec_ref_t)
```
uint32_t field = 3;
uint32_t a_regular_array[10] = {...};
//flatbuffers_uint32_vec_ref_t array;
//array = flatbuffers_uint32_vec_create(&builder, a_regular_array, 10);
uint32_t array[10];

ns_Payload_union_ref_t pl = ns_Payload_as_Struct2(ns_Struct2_create(&builder, field, array));
ns_MyTable_ref_t tbl = Cfg_Configuration_create(&builder, pl);
```

**Reply by mikkelfj** [reply]
Great, I'm glad it worked out. Thanks for letting me know.

**Comment by mikkelfj** [comment, marked as answer]
It is also possible your first example might work if you replace the vector reference with a native C array of the correct type, and leave the creation of the vector.

In the documentation, you will see an assign call that takes an array and string as arguments. It should work the same with create calls.
https://github.com/dvidelabs/flatcc/blob/master/doc/builder.md#fixed-length-arrays-in-structs
